torch/distributed/pipelining/schedules.py

```diff
@@ -1394,3 +1498,64 @@ def backward_stage_index(step):
             forward_stage_index,
             backward_stage_index,
         )
+
+    def _add_bubbles_to_actions(self, num_stages_global):
```

Comment:
I like the idea of this function. If I'm understanding correctly this logic is meant to prevent actions with dependencies from each other from being fired before their dependent action (e.g. Backward_mb_1_stage_3 being called before backward_mb_1_stage_4).

I guess we were preventing this in interleaved1f1b by appending `None` at some parts during the action order generation, but now we are adding this post-generation for zero-bubble? Is this because the logic is too hard to encode correctly during generation?

Reply:
My thinking was that a sanity check and cleanup function at the end might be easier and cleaner than a bunch of if elses in the actual logic itself (e.g. for cases where mb < stages), but if you feel strongly that we should do everything in place I can think about moving this there too.